src/Markup/Perspex.Markup/Data/Parsers/LiteralParser.cs

```diff
@@ -12,21 +12,33 @@ public static object Parse(Reader r)
         {
             if (char.IsDigit(r.Peek))
             {
-                StringBuilder result = new StringBuilder();
-
+                var result = new StringBuilder();
+                var foundDecimal = false;
                 while (!r.End)
                 {
                     if (char.IsDigit(r.Peek))
                     {
                         result.Append(r.Take());
                     }
+                    else if (!foundDecimal && r.Peek == '.')
+                    {
+                        result.Append(r.Take());
+                        foundDecimal = true;
+                    }
                     else
                     {
                         break;
                     }
                 }
 
-                return int.Parse(result.ToString(), CultureInfo.InvariantCulture);
+                if (!foundDecimal)
+                {
+                    return int.Parse(result.ToString(), CultureInfo.InvariantCulture); 
+                }
+                else
+                {
+                    return result.ToString(); // Leave as a string to support double, float, and decimal indicies
```

Comment:
I'm up for doing the conversions later. However, since the conversions are done at bind-time, not at construction, it makes it difficult to have the checking for `IList` and `Array` that is already there and the optimized `INotifyCollectionChanged` (which I think currently might have problems with non-integer arguments).

We could just special case them again afterwards, just swallow the possible exceptions from invalid indicies and return `UnsetValue`, or just let the exceptions bubble up to the user since they put in the bad binding.